File: 12-27-2024/Search_A_2D_Matrix.java
package practice;

public class Search_A_2D_Matrix {
    public static boolean searchMatrix(int[][] matrix, int target) {
        int n = matrix.length;
        int m = matrix[0].length;

        int i = 0, j = n*m-1;

        while(i < j) {
            int mid = (i + j) / 2;

            int x = mid / m, y = mid % m;
            if(matrix[x][y] == target) return true;

            if(matrix[x][y] > target)
                j = mid;
            else
                i = mid + 1;
        }

        return matrix[i / m][i % m] == target;
    }
    public static void main(String[] args) {
        int[][] arr = new int[][]{{1,1}};
        System.out.println(searchMatrix(arr, 2));
    }
}

File: 2026-02-08/Search_A_2D_Matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix)-1, len(matrix[0])-1
        l, r = 0, m
        while l <= r:
            mid = l + (r-l) // 2
            if matrix[mid][n] == target:
                return True
            elif target > matrix[mid][n]:
                l = mid + 1
            else:
                r = mid - 1

        if l > m:
            return False

        row = l
        l, r = 0, n
        while l <= r:
            mid = l + (r - l) // 2
            if matrix[row][mid] == target:
                return True
            elif target > matrix[row][mid]:
                l = mid + 1
            else:
                r = mid - 1

        return False


sol = Solution()
print(sol.searchMatrix([[1]], 2))
